library/src/main/java/kwasm/format/text/Type.kt

```diff
@@ -59,39 +61,107 @@ sealed class Type<T>(
+
+                            if (internalSequence.contains("(param") && parsingParams) {
+                                params.add(Param(internalSequence, context).value)
+                            } else if (internalSequence.contains("(result")) {
+                                parsingParams = false
+                                returnValues.add(Result(internalSequence, context).value)
+                            } else {
+                                throw ParseException("Invalid FunctionType Syntax", context)
+                            }
+                            internalSequence = ""
+                        }
+                        else -> {
+                            internalSequence += "$token "
+                        }
+                    }
+                }
+                return FunctionType(params, returnValues)
+            }
         }
     }
 
+    /**
+     * Encapsulates kwasm.ast.Param for parsing. Parses the sequence in accordance
+     * to the definition in [the docs](https://webassembly.github.io/spec/core/text/types.html#value-types)
+     *
+     * @constructor Parses the sequence passed in and populates the appropriate parsed value.
+     * @throws ParseException when the sequence falls outside of the spec definition
+     */
     class Param(
         sequence: CharSequence,
         context: ParseContext? = null
-    ) : Type<Unit>(sequence, context) {
-        override fun parseValue() {
-            TODO("not implemented")
+    ) : Type<kwasm.ast.Param>(sequence, context) {
+        override fun parseValue(): kwasm.ast.Param {
+            val keywordAndParameters = getOperationAndParameters(sequence, context)
+            if (keywordAndParameters.first != "param" || keywordAndParameters.second.size > 2) {
+                throw ParseException("Invalid Param syntax", context.shiftColumnBy(1))
+            }
+            return if (keywordAndParameters.second.size == 1) {
+                // Context is shifted by 7 to shift past '(param ' to the start of the actual ValueType
+                kwasm.ast.Param(null, ValueType(keywordAndParameters.second[0], context.shiftColumnBy(7)).value)
+            } else {
+                // Context is shifted by 8 + id length to shift past '(param id ' to the start of the actual ValueType
+                kwasm.ast.Param(
+                    Identifier.Local(keywordAndParameters.second[0].toString()),
+                    ValueType(
+                        keywordAndParameters.second[1],
+                        context.shiftColumnBy(8 + keywordAndParameters.second[0].length)
+                    ).value
+                )
+            }
         }
     }
 
     class Result(
         sequence: CharSequence,
         context: ParseContext? = null
-    ) : Type<Unit>(sequence, context) {
-        override fun parseValue() {
-            TODO("not implemented")
+    ) : Type<kwasm.ast.ValueType>(sequence, context) {
```

Comment:
Should this be `Result`, instead of `ValueType`?

Reply:
Are you suggesting creating another data class?

Reply:
I think if `Result` is its own type in the spec, it should probably be its own AST node.